File: src/jzrlib/core/Directions.java
/*
 *  "JZRLib", Java Roguelike games development Library.
 *  Copyright (C) 2015 by Serg V. Zhdanovskih (aka Alchemist, aka Norseman).
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package jzrlib.core;

/**
 *
 * @author Serg V. Zhdanovskih
 */
public final class Directions extends jzrlib.core.FlagSet
{
    public static final int dtNone = 0;

    public static final int dtNorth = 1;
    public static final int dtSouth = 2;
    public static final int dtWest = 3;
    public static final int dtEast = 4;

    public static final int dtNorthWest = 5;
    public static final int dtNorthEast = 6;
    public static final int dtSouthWest = 7;
    public static final int dtSouthEast = 8;

    public static final int dtZenith = 9;
    public static final int dtNadir = 10;
    public static final int dtPlace = 11; // TODO: checkit and remove


    public static final int dtFirst = 1;
    public static final int dtLast = 4;
    public static final int dtFlatFirst = 1;
    public static final int dtFlatLast = 8;

    public Directions(int... args)
    {
        super(args);
    }

    public Directions(Directions set)
    {
        super(set);
    }

    public static final DirectionRec[] Data;
    private static final int[] Idx2Dir;
    public static int[] IsoTrans;

    static {
        Data = new DirectionRec[12];
        Data[0] = new DirectionRec(0, 0, 0, Directions.dtNone, '?', '?');
        Data[1] = new DirectionRec(0, -1, 0, Directions.dtSouth, 'N', 'S');
        Data[2] = new DirectionRec(0, 1, 0, Directions.dtNorth, 'S', 'N');
        Data[3] = new DirectionRec(-1, 0, 0, Directions.dtEast, 'W', 'E');
        Data[4] = new DirectionRec(1, 0, 0, Directions.dtWest, 'E', 'W');
        Data[5] = new DirectionRec(-1, -1, 0, Directions.dtSouthEast, '\0', '\0');
        Data[6] = new DirectionRec(1, -1, 0, Directions.dtSouthWest, '\0', '\0');
        Data[7] = new DirectionRec(-1, 1, 0, Directions.dtNorthEast, '\0', '\0');
        Data[8] = new DirectionRec(1, 1, 0, Directions.dtNorthWest, '\0', '\0');
        Data[9] = new DirectionRec(0, 0, -1, Directions.dtNadir, '\0', '\0');
        Data[10] = new DirectionRec(0, 0, 1, Directions.dtZenith, '\0', '\0');
        Data[11] = new DirectionRec(0, 0, 0, Directions.dtNone, '\0', '\0');

        Idx2Dir = new int[]{
            Directions.dtNone, 
            Directions.dtSouthWest, 
            Directions.dtSouth, 
            Directions.dtSouthEast, 
            Directions.dtWest, 
            Directions.dtPlace, 
            Directions.dtEast, 
            Directions.dtNorthWest, 
            Directions.dtNorth, 
            Directions.dtNorthEast};

        IsoTrans = new int[]{Directions.dtNone, Directions.dtNorthWest, Directions.dtSouthEast, Directions.dtSouthWest, Directions.dtNorthEast, Directions.dtWest, Directions.dtNorth, Directions.dtSouth, Directions.dtEast, Directions.dtZenith, Directions.dtNadir, Directions.dtPlace};
    }

    public static final class DirectionRec
    {
        public int dX;
        public int dY;
        public int dZ;
        public int Opposite;
        public char symDirect;
        public char symOpposite;

        public DirectionRec(int dX, int dY, int dZ, int opposite, char symDirect, char symOpposite)
        {
            this.dX = dX;
            this.dY = dY;
            this.dZ = dZ;
            this.Opposite = opposite;
            this.symDirect = symDirect;
            this.symOpposite = symOpposite;
        }
    }

    /**
     * Returns direction of movements by changes of coordinates.
     * @param pX Old x-coord
     * @param pY Old y-coord
     * @param newX New x-coord
     * @param newY New y-coord
     * @return Direction of movement
     */
    public static int getDirByCoords(int pX, int pY, int newX, int newY)
    {
        int dx = Integer.signum(newX - pX);
        int dy = Integer.signum(newY - pY);
        int dir = (8 + dx - (dy + 1) * 3);
        return Idx2Dir[dir];
    }
}

File: project/Core/Directions.cs
/*
 *  "ZRLib", Roguelike games development Library.
 *  Copyright (C) 2015 by Serg V. Zhdanovskih.
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

using System;

namespace ZRLib.Core
{
    public sealed class Directions : FlagSet
    {
        public const int DtNone = 0;

        public const int DtNorth = 1;
        public const int DtSouth = 2;
        public const int DtWest = 3;
        public const int DtEast = 4;

        public const int DtNorthWest = 5;
        public const int DtNorthEast = 6;
        public const int DtSouthWest = 7;
        public const int DtSouthEast = 8;

        public const int DtZenith = 9;
        public const int DtNadir = 10;
        public const int DtPlace = 11;
        // TODO: checkit and remove


        public const int DtFirst = 1;
        public const int DtLast = 4;
        public const int DtFlatFirst = 1;
        public const int DtFlatLast = 8;

        public Directions(params int[] args)
            : base(args)
        {
        }

        public Directions(Directions set)
            : base(set)
        {
        }

        public static readonly DirectionRec[] Data;
        private static readonly int[] Idx2Dir;
        public static int[] IsoTrans;

        static Directions()
        {
            Data = new DirectionRec[12];
            Data[0] = new DirectionRec(0, 0, 0, Directions.DtNone, '?', '?');
            Data[1] = new DirectionRec(0, -1, 0, Directions.DtSouth, 'N', 'S');
            Data[2] = new DirectionRec(0, 1, 0, Directions.DtNorth, 'S', 'N');
            Data[3] = new DirectionRec(-1, 0, 0, Directions.DtEast, 'W', 'E');
            Data[4] = new DirectionRec(1, 0, 0, Directions.DtWest, 'E', 'W');
            Data[5] = new DirectionRec(-1, -1, 0, Directions.DtSouthEast, '\0', '\0');
            Data[6] = new DirectionRec(1, -1, 0, Directions.DtSouthWest, '\0', '\0');
            Data[7] = new DirectionRec(-1, 1, 0, Directions.DtNorthEast, '\0', '\0');
            Data[8] = new DirectionRec(1, 1, 0, Directions.DtNorthWest, '\0', '\0');
            Data[9] = new DirectionRec(0, 0, -1, Directions.DtNadir, '\0', '\0');
            Data[10] = new DirectionRec(0, 0, 1, Directions.DtZenith, '\0', '\0');
            Data[11] = new DirectionRec(0, 0, 0, Directions.DtNone, '\0', '\0');

            Idx2Dir = new int[] {
                Directions.DtNone,
                Directions.DtSouthWest,
                Directions.DtSouth,
                Directions.DtSouthEast,
                Directions.DtWest,
                Directions.DtPlace,
                Directions.DtEast,
                Directions.DtNorthWest,
                Directions.DtNorth,
                Directions.DtNorthEast
            };

            IsoTrans = new int[] {
                Directions.DtNone,
                Directions.DtNorthWest,
                Directions.DtSouthEast,
                Directions.DtSouthWest,
                Directions.DtNorthEast,
                Directions.DtWest,
                Directions.DtNorth,
                Directions.DtSouth,
                Directions.DtEast,
                Directions.DtZenith,
                Directions.DtNadir,
                Directions.DtPlace
            };
        }

        public sealed class DirectionRec
        {
            public int DX;
            public int DY;
            public int DZ;
            public int Opposite;
            public char SymDirect;
            public char SymOpposite;

            public DirectionRec(int dX, int dY, int dZ, int opposite, char symDirect, char symOpposite)
            {
                DX = dX;
                DY = dY;
                DZ = dZ;
                Opposite = opposite;
                SymDirect = symDirect;
                SymOpposite = symOpposite;
            }
        }

        /// <summary>
        /// Returns direction of movements by changes of coordinates. </summary>
        /// <param name="pX"> Old x-coord </param>
        /// <param name="pY"> Old y-coord </param>
        /// <param name="newX"> New x-coord </param>
        /// <param name="newY"> New y-coord </param>
        /// <returns> Direction of movement </returns>
        public static int GetDirByCoords(int pX, int pY, int newX, int newY)
        {
            int dx = Math.Sign(newX - pX);
            int dy = Math.Sign(newY - pY);
            int dir = (8 + dx - (dy + 1) * 3);
            return Idx2Dir[dir];
        }
    }
}
